File: src/go/4_BehavioralPatterns/6_Memento/undoredo.go
package memento

import "fmt"

type Memento struct {
  Balance int
}

type BankAccount struct {
  balance int
  changes []*Memento
  current int
}

func (b *BankAccount) String() string {
  return fmt.Sprint("Balance = $", b.balance,
    ", current = ", b.current)
}

func NewBankAccount(balance int) *BankAccount {
  b := &BankAccount{balance: balance}
  b.changes = append(b.changes, &Memento{balance})
  return b
}

func (b *BankAccount) Deposit(amount int) *Memento {
  b.balance += amount
  m := Memento{b.balance}
  b.changes = append(b.changes, &m)
  b.current++
  fmt.Println("Deposited", amount,
    ", balance is now", b.balance)
  return &m
}

func (b *BankAccount) Restore(m *Memento) {
  if m != nil {
    b.balance -= m.Balance
    b.changes = append(b.changes, m)
    b.current = len(b.changes) - 1
  }
}

func (b *BankAccount) Undo() *Memento {
  if b.current > 0 {
    b.current--
    m := b.changes[b.current]
    b.balance = m.Balance
    return m
  }
  return nil // nothing to undo
}

func (b *BankAccount) Redo() *Memento {
  if b.current + 1 < len(b.changes) {
    b.current++
    m := b.changes[b.current]
    b.balance = m.Balance
    return m
  }
  return nil
}

func main() {
  ba := NewBankAccount(100)
  ba.Deposit(50)
  ba.Deposit(25)
  fmt.Println(ba)

  ba.Undo()
  fmt.Println("Undo 1:", ba)
  ba.Undo()
  fmt.Println("Undo 2:", ba)
  ba.Redo()
  fmt.Println("Redo:", ba)
}

File: src/swift/4_BehavioralPatterns/6_Memento/UndoRedo.swift
import Foundation

class Memento
{
  let balance: Int
  init(_ balance: Int)
  {
    self.balance = balance
  }
}

class BankAccount : CustomStringConvertible
{
  private var balance: Int
  private var changes = [Memento]()
  private var current = 0

  init(_ balance: Int)
  {
    self.balance = balance
    changes.append(Memento(balance))
  }

  func deposit(_ amount: Int) -> Memento
  {
    balance += amount
    let m = Memento(balance)
    changes.append(m)
    current += 1
    return m
  }

  func restore(_ m: Memento?)
  {
    if let mm = m
    {
      balance = mm.balance
      changes.append(mm)
      current = changes.count - 1
    }
  }

  func undo() -> Memento?
  {
    if current > 0
    {
      current -= 1
      let m = changes[current]
      balance = m.balance
      return m
    }
    return nil
  }

  func redo() -> Memento?
  {
    if (current+1) < changes.count
    {
      current += 1
      let m = changes[current]
      balance = m.balance
      return m
    }
    return nil
  }

  public var description : String
  {
    return "Balance = \(balance)"
  }
}

func main()
{
  var ba = BankAccount(100)
  ba.deposit(50)
  ba.deposit(25)
  print(ba)

  ba.undo()
  print("Undo 1: \(ba)")
  ba.undo()
  print("Undo 2: \(ba)")
  ba.redo()
  print("Redo 2: \(ba)")
}

main()